compiler/rustc_mir/src/dataflow/impls/available_locals.rs

```diff
@@ -0,0 +1,338 @@
+    }
+
+    /// Returns the state with context attached, making debug easier
+    pub fn debug_state(&'a self, state: &'a State) -> impl fmt::Debug + 'a {
+        DebugWithAdapter { this: state, ctxt: &self.local_with_location_map }
+    }
+
+    /// Returns whether the given local is available at the given state
+    pub fn is_available(&self, local: Local, state: &'a State) -> bool {
+        self.local_with_location_map.local_with_locations(local).any(
+            |place_taken_ref_local_with_location| {
+                state.0.contains(place_taken_ref_local_with_location)
+            },
+        )
+    }
+
+    fn transfer_function<T>(&'a self, available: &'a mut T) -> TransferFunction<'a, T>
+    where
+        T: GenKill<LocalWithLocationIndex>,
+    {
+        TransferFunction { available, local_with_location_map: &self.local_with_location_map }
+    }
+}
+impl AnalysisDomain<'tcx> for AvailableLocals {
+    /// We want a *forward must* analysis.
+    /// At each join point, we only keep locals that were available in both basic blocks.
+    /// For this, we use the Dual of a semi-lattice, such that intersection is the join operator
+    type Domain = State;
+    type Direction = Forward;
+
+    const NAME: &'static str = "available_locals";
+
+    fn bottom_value(&self, _: &mir::Body<'tcx>) -> Self::Domain {
+        // If we were not using the dual, we would have nothing available, but since we are using
+        // the dual, we also "reverse" the bottom to be filled
+        // bottom = nothing is available
+        let bottom = BitSet::new_filled(self.local_with_location_map.len());
+        lattice::Dual(bottom)
+    }
+
+    fn initialize_start_block(&self, body: &mir::Body<'tcx>, state: &mut Self::Domain) {
+        // In the start block, nothing is available at first
+        state.0.clear();
+
+        // only the function parameters are available
+        for arg in body.args_iter() {
+            state.0.insert(self.local_with_location_map.local_with_specific_location(arg, None));
+        }
+    }
+}
+
+impl GenKillAnalysis<'tcx> for AvailableLocals {
+    type Idx = LocalWithLocationIndex;
+
+    fn statement_effect(
+        &self,
+        trans: &mut impl GenKill<Self::Idx>,
+        statement: &mir::Statement<'tcx>,
+        location: Location,
+    ) {
+        self.transfer_function(trans).visit_statement(statement, location);
+    }
+
+    fn terminator_effect(
+        &self,
+        trans: &mut impl GenKill<Self::Idx>,
+        terminator: &mir::Terminator<'tcx>,
+        location: Location,
+    ) {
+        self.transfer_function(trans).visit_terminator(terminator, location);
+    }
+
+    fn call_return_effect(
+        &self,
+        _trans: &mut impl GenKill<Self::Idx>,
+        _block: mir::BasicBlock,
+        _func: &mir::Operand<'tcx>,
+        _args: &[mir::Operand<'tcx>],
+        _dest_place: mir::Place<'tcx>,
+    ) {
+        // Conservatively do not try to reason about calls
```

Comment:
We should take care here to invalidate operands that move locals, i'll fix this in a commit